hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/snapshot/OMSnapshotMoveTableKeysRequest.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.ratis.server.protocol.TermIndex;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.hadoop.hdds.HddsUtils.fromProtobuf;
+import static org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature.FILESYSTEM_SNAPSHOT;
+
+/**
+ * Handles OMSnapshotMoveTableKeysRequest Request.
+ * This is an OM internal request. Does not need @RequireSnapshotFeatureState.
+ */
+public class OMSnapshotMoveTableKeysRequest extends OMClientRequest {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OMSnapshotMoveTableKeysRequest.class);
+
+  public OMSnapshotMoveTableKeysRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  @DisallowedUntilLayoutVersion(FILESYSTEM_SNAPSHOT)
+  public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager, TermIndex termIndex) {
+    OmMetadataManagerImpl omMetadataManager = (OmMetadataManagerImpl) ozoneManager.getMetadataManager();
+    SnapshotChainManager snapshotChainManager = omMetadataManager.getSnapshotChainManager();
+
+    SnapshotMoveTableKeysRequest moveTableKeysRequest = getOmRequest().getSnapshotMoveTableKeysRequest();
+
+    OMClientResponse omClientResponse;
+    OzoneManagerProtocolProtos.OMResponse.Builder omResponse = OmResponseUtil.getOMResponseBuilder(getOmRequest());
+    try {
+      SnapshotInfo fromSnapshot = SnapshotUtils.getSnapshotInfo(ozoneManager,
+          snapshotChainManager, fromProtobuf(moveTableKeysRequest.getFromSnapshotID()));
+      // If there is no Non-Deleted Snapshot move the
+      // keys to Active Object Store.
+      SnapshotInfo nextSnapshot = SnapshotUtils.getNextSnapshot(ozoneManager, snapshotChainManager, fromSnapshot);
+
+      // If next snapshot is not active then ignore move. Since this could be a redundant operations.
+      if (nextSnapshot != null && nextSnapshot.getSnapshotStatus() != SnapshotInfo.SnapshotStatus.SNAPSHOT_ACTIVE) {
+        throw new OMException("Next snapshot : " + nextSnapshot + " in chain is not active.",
+            OMException.ResultCodes.INVALID_SNAPSHOT_ERROR);
+      }
+
+      String expectedBucketKeyPrefix = omMetadataManager.getBucketKeyPrefix(fromSnapshot.getVolumeName(),
+          fromSnapshot.getBucketName());
+      String expectedBucketKeyPrefixFSO = omMetadataManager.getBucketKeyPrefixFSO(fromSnapshot.getVolumeName(),
+          fromSnapshot.getBucketName());
+
+      // Filter only deleted keys with atlest one keyInfo per key.
+      Set<String> keys = new HashSet<>();
+      List<SnapshotMoveKeyInfos> deletedKeys =
+          moveTableKeysRequest.getDeletedKeysList().stream()
+              .filter(snapshotMoveKeyInfos -> !snapshotMoveKeyInfos.getKeyInfosList().isEmpty())
+              .collect(Collectors.toList());
```

Comment:
This is an extra traversal. Anyways you are traversing it in loop 101-112.

Same for other rename and dir.

Reply:
+1 on this, filter is doing a traversal and we are also doing a traversal below. We should do in the latter traversal itself.

Reply:
Just realized we can move this to pre execute and this shouldn't be done in validateAndUpdateCache.

Reply:
We don't need to send it over ratis to do these verifications. We can even do prefix verification since we are not going to support bucket rename & volume rename operation anytime soon.